specialcouscous/rf_parallel.py

```
@@ -311,47 +363,74 @@ def train(
             # Array with one vector for each tree in global RF with predictions for all test samples.
             # Final prediction of parallel RF is majority vote over all sub estimators.
             # Calculate majority vote.
             log.info(f"[{rank}/{size}]: Calculate majority vote.")
             majority_vote = self._calc_majority_vote(tree_predictions)
-        else:
+            # Calculate accuracy of global model over all local datasets.
+            n_correct_local = (targets == majority_vote).sum()
+            n_samples_local = targets.shape[0]
+            log.info(
+                f"[{rank}/{size}]: Fraction of correctly predicted samples on this rank: "
+                f"{n_correct_local} / {n_samples_local}."
+            )
+            n_correct = self.comm.allreduce(n_correct_local)
+            n_samples = self.comm.allreduce(targets.shape[0])
+            log.info(
+                f"[{rank}/{size}]: Fraction of correctly predicted samples overall:"
+                f" {n_correct}/ {n_samples}"
+            )
+            # Calculate accuracy of global model on global dataset.
+            self.acc_global = n_correct / n_samples
+            # Calculate accuracy of global model on local dataset.
+            # Note that this metric can only be calculated if the global model is shared among all ranks.
+            self.acc_global_local = (targets == majority_vote).mean()
+
+        else:  # Global model is not shared. Note that the dataset to be tested must be shared among all ranks.
             # Get class predictions of sub estimators in each forest.
             log.info(f"[{rank}/{size}]: Get predictions of individual sub estimators.")
-            tree_predictions_local = self._predict_locally(test_samples)
+            tree_predictions_local = self._predict_locally(samples)
             log.info(f"[{rank}/{size}]: Calculate majority vote via histograms.")
             predicted_class_hists = self._predicted_class_hist(
```

Should we make sure the the samples have the same length here so we can raise an Exception and exit more gracefully instead of failing during the communication?

---

You mean allgathering the local lenghts of the samples and checking whether they are the same before continuing? We can do that as a sanity check but that would not fully prevent "misuse" of this function as it would still work technically if the local datasets to run evalation on have the same number of samples but the samples itself are different. That could, e.g., occur when only distributing the dataset over the ranks without imbalance in the case `num_samples % num_ranks` is 0.

---

Exactly, it would be primarily to catch the truncated message errors and give a more useful error message in that case.

---

Done.